Storage/Storage/SQL/BrowserDB.swift

```diff
@@ -53,57 +40,95 @@ class BrowserDB {
     // XXX: Increasing this should blow away old history, since we currently dont' support any upgrades
     private let Version: Int = 3
     private let FileName = "browser.db"
+    private let files: FileAccessor
+    private let tableTable: TableTable<TableInfo>
 
     private var initialized = [String]()
 
-    private func exists<T : Table>(db: SQLiteDBConnection, table: T) -> Bool {
-        var found = false
-        let sqlStr = "SELECT name FROM sqlite_master WHERE type = 'table' AND name=?"
-        let res = db.executeQuery(sqlStr, factory: StringFactory, withArgs: [table.name])
-        return res.count > 0
-    }
-
     init?(files: FileAccessor) {
+        self.files = files
         db = SwiftData(filename: files.get(FileName)!)
+        self.tableTable = TableTable()
+        self.createOrUpdate(self.tableTable)
+    }
+
+    var filename: String {
+        return db.filename
+    }
+
+    // Creates a table and writes its table info the the table-table database.
+    private func createTable<T:Table>(db: SQLiteDBConnection, table: T) -> Bool {
+        debug("Try create \(table.name) version \(table.version)")
+        if !table.create(db, version: table.version) {
+            // If creating failed, we'll bail without storing the table info
+            return false
+        }
+
+        var err: NSError? = nil
+        return tableTable.insert(db, item: table, err: &err) > -1
+    }
+
+    // Updates a table and writes its table info the the table-table database.
+    private func updateTable<T: Table>(db: SQLiteDBConnection, table: T) -> Bool {
+        debug("Try update \(table.name) version \(table.version)")
+        var from = 0
+        // Try to find the stored version of the table
+        let cursor = tableTable.query(db, options: QueryOptions(filter: table.name))
```

Comment:
Side comment: are we confident in when our cursors' `deinit` method will be called? I wonder if we ought to have explicit closure.